search-provider/eks-article-metadata-query-provider.c

```diff
@@ -0,0 +1,428 @@
+                                                      &builder,
+                                                      "tags",
+                                                      G_VARIANT_TYPE ("as"));
+
+      g_variant_builder_close (&builder);
+    }
+
+  shards_strv = strv_from_shard_list ((GSList *) shards);
+
+  eks_article_metadata_complete_query (state->provider->skeleton,
+                                       state->invocation,
+                                       (const gchar * const *) shards_strv,
+                                       g_variant_builder_end (&builder));
+}
+
+static void
+append_construction_prop_from_string (const gchar *key,
+                                      const gchar *str,
+                                      GArray      *values_array,
+                                      GPtrArray   *props_array,
+                                      guint       *nprops)
+{
+  guint index = values_array->len;
+  GValue *value = NULL;
+
+  /* Need to be careful to set value *after* the array has been resized */
+  g_array_set_size (values_array, values_array->len + 1);
+  value = &(g_array_index (values_array, GValue, index));
+  g_value_init (value, G_TYPE_STRING);
+  g_value_set_string (value, str);
+  g_ptr_array_add (props_array, g_strdup (key));
+  ++(*nprops);
+}
+
+static void
+append_construction_prop_from_variant (const gchar *key,
+                                       GVariant    *variant,
+                                       GArray      *values_array,
+                                       GPtrArray   *props_array,
+                                       guint       *nprops)
+{
+  guint index = values_array->len;
+  GValue *value = NULL;
+
+  /* Need to be careful to set value *after* the array has been resized */
+  g_array_set_size (values_array, values_array->len + 1);
+  value = &(g_array_index (values_array, GValue, index));
+  g_dbus_gvariant_to_gvalue (variant, value);
+  g_ptr_array_add (props_array, g_strdup (key));
+  ++(*nprops);
+}
+
+static EkncQueryObject *
+create_query_from_dbus_query_parameters (GVariant    *query_parameters,
+                                         const gchar *application_id)
+{
+  GVariantIter iter;
+  gchar *iter_key;
+  GVariant *iter_value;
+
+  g_autoptr(GArray) values_array = g_array_new (FALSE, TRUE, sizeof (GValue));
+  g_autoptr(GPtrArray) props_array = g_ptr_array_new_with_free_func (g_free);
+  guint nprops = 0;
+
+  /* Always add the app-id to the query */
+  append_construction_prop_from_string ("app-id",
+                                        application_id,
+                                        values_array,
+                                        props_array,
+                                        &nprops);
+
+  g_variant_iter_init (&iter, query_parameters);
+  while (g_variant_iter_next (&iter, "{sv}", &iter_key, &iter_value))
+    {
+      g_autofree gchar *key = iter_key;
+      g_autoptr(GVariant) variant = iter_value;
```

Comment:
Why not just do
```
append_construction_prop_from_variant (key, variant, ...);
```

Reply:
As opposed to checking the key-name each time? I suppose that could work for now. I think the original reason why I had the if statements is in case I needed to do any translation between key names but that's not the case here. In any event, I think it is still prudent to check that that the key name is valid before adding the construction prop to avoid setting invalid properties on the object.

Reply:
done in 2728c3e

Reply:
Yeah, that's what I meant, even though I have a follow up comment.